cdap-formats/src/main/java/co/cask/cdap/format/DelimitedStringsRecordFormat.java

```diff
@@ -114,15 +134,120 @@ private boolean isStringArray(Schema schema) {
+  }
+
+  /**
+   * {@link RecordMaker} that uses the default schema.
+   */
+  private static class DefaultRecordMaker implements RecordMaker {
+
+    @Override
+    public StructuredRecord make(Schema schema, Iterator<String> bodyFields) {
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+      List<String> fields = Lists.newArrayList(bodyFields);
+      builder.set("body", fields.toArray(new String[fields.size()]));
+
+      return builder.build();
+    }
+  }
+
+  /**
+   * {@link RecordMaker} that uses a schema.
+   */
+  private static class SchemaRecordMaker implements RecordMaker {
+
+    @Override
+    public StructuredRecord make(Schema schema, Iterator<String> bodyFields) {
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+      Iterator<Schema.Field> fieldsIterator = schema.getFields().iterator();
+      while (fieldsIterator.hasNext()) {
+        Schema.Field field = fieldsIterator.next();
+        Schema fieldSchema = field.getSchema();
+        String fieldName = field.getName();
+        if (isStringArray(fieldSchema)) {
+          if (!fieldsIterator.hasNext()) {
+            // only do varargs-style string array parsing on bodyField if it's the last field
+            List<String> fields = Lists.newArrayList(bodyFields);
+            builder.set(fieldName, fields.toArray(new String[fields.size()]));
+          } else {
+            throw new UnexpectedFormatException(
+              String.format("string array type field '%s' must be the last schema field", fieldName));
+          }
+        } else {
+          // simple type (not string array)
+          String bodyField = bodyFields.hasNext() ? bodyFields.next() : null;
+          String val = parseBodyValue(bodyField, fieldSchema);
+          builder.convertAndSet(fieldName, val);
+        }
+      }
+      return builder.build();
+    }
+  }
+
+  /**
+   * {@link RecordMaker} that uses the "mapping" setting and a schema.
+   */
+  private static class MappedSchemaRecordMaker implements RecordMaker {
+    private final Map<String, Integer> mapping;
+    private final int lastMappingIndex;
+
+    private MappedSchemaRecordMaker(Map<String, Integer> mapping) {
+      this.mapping = mapping;
+      this.lastMappingIndex = Collections.max(mapping.values());
+    }
+
+    @Override
+    public StructuredRecord make(Schema schema, Iterator<String> bodyFields) {
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+      // TODO: only read what's necessary from event.getBody() (e.g. if mapping is "0:f0", then only read first entry)
+      List<String> fields = Lists.newArrayList(Iterators.limit(bodyFields, lastMappingIndex + 1));
+      for (Schema.Field field : schema.getFields()) {
+        Schema fieldSchema = field.getSchema();
+        String fieldName = field.getName();
+        int index = mapping.get(fieldName);
+        String val = parseBodyValue(fields.get(index), fieldSchema);
```

Comment:
Fixed in https://github.com/caskdata/cdap/commit/7deea63cd488e81b155a1da8c9f3180c42625e14